src/mesh/coordinates.cxx

```diff
@@ -170,6 +170,129 @@ Coordinates::Coordinates(Mesh *mesh)
   }
 }
 
+// Interpolate a Field2D to a new CELL_LOC with interp_to.
+// Communicates to set internal guard cells.
+// Boundary guard cells are set equal to the nearest grid point (equivalent to
+// 2nd order accurate Neumann boundary condition).
+// Corner guard cells are set to garbage values, because that is the simplest
+// way to allow error checking of the rest of the guard cells in a single loop,
+// without the corner points always failing.
+Field2D interpolate_and_Neumann(const Field2D &f, CELL_LOC location, bool diagonal) {
+  Mesh* localmesh = f.getMesh();
+  Field2D result = interp_to(f, location, RGN_NOBNDRY);
+  localmesh->communicate(result);
+
+  // Copy nearest value into boundaries so that differential geometry terms can
+  // be interpolated if necessary
+  // Note: cannot use applyBoundary("neumann") here because applyBoundary()
+  // would try to create a new Coordinates object since we have not finished
+  // initializing yet, leading to an infinite recursion
+  for (auto bndry : localmesh->getBoundaries()) {
+    if (bndry->bx != 0) {
+      // If bx!=0 we are on an x-boundary, inner if bx>0 and outer if bx<0
+      for(bndry->first(); !bndry->isDone(); bndry->next1d()) {
+        for (int i=0; i<localmesh->xstart; i++)
+          result(bndry->x+i*bndry->bx,bndry->y) = result(bndry->x+(i-1)*bndry->bx, bndry->y-bndry->by);
+      }
+    }
+    if (bndry->by != 0) {
+      // If by!=0 we are on a y-boundary, upper if by>0 and lower if by<0
+      for(bndry->first(); !bndry->isDone(); bndry->next1d()) {
+        for (int i=0; i<localmesh->ystart; i++)
+          result(bndry->x,bndry->y+i*bndry->by) = result(bndry->x-bndry->bx, bndry->y+(i-1)*bndry->by);
+      }
+    }
+  }
+
+  // Set corner guard cells to avoid exceptions
+  // Set diagonal elements to a large value, and off-diagonal to zero
+  if (diagonal) {
+    for (int i=0; i<localmesh->xstart; i++) {
+      for (int j=0; j<localmesh->ystart; j++) {
+        result(i, j) = 1.e4;
```

Comment:
I think I had issues at the time when checkData was defaulting to RGN_ALL. Seems not to be a problem now, so I've set all the corner guard cells to BoutNaN.

Reply:
Although now the code does call `invert3x3` with BoutNaNs as input, in a loop over all points
https://github.com/boutproject/BOUT-dev/blob/4547f8c1b264bf9820bec1c30c9fdc9a846193b1/src/mesh/coordinates.cxx#L553-L557
This 'works' because the condition for the error check in `invert3x3` will evaluate to `false` with NaNs (I guess)
https://github.com/boutproject/BOUT-dev/blob/4547f8c1b264bf9820bec1c30c9fdc9a846193b1/include/utils.hxx#L230-L236
but this seems a bit dirty... Would probably be better to loop over just grid cells and non-corner guard cells. But maybe that can wait until we have new `REGION` goodness #1245.

Reply:
...I think it turns out I didn't test this properly until I merged it into `set_coordinates_location` where the non-`CELL_CENTRE` coordinates actually get used: with NaN in the corner guard cells, the interpolated metrics fail the check that they are finite
https://github.com/boutproject/BOUT-dev/blob/99edc6ed68906ed0d7ee0164f0947a93db1d0d6e/src/mesh/coordinates.cxx#L249-L252
I think the easiest way to get around this is to use a region argument of `finite()` to skip the guard cells.